transport/http/middleware_headers.go

```diff
@@ -0,0 +1,91 @@
+package http
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+
+	"github.com/awslabs/smithy-go/middleware"
+)
+
+var randReader = rand.Reader
+
+type headerValue struct {
+	header string
+	value  string
+	append bool
+}
+
+type headerValueHelper struct {
+	headerValues []headerValue
+}
+
+func (h *headerValueHelper) addHeaderValue(value headerValue) {
+	h.headerValues = append(h.headerValues, value)
+}
+
+func (h *headerValueHelper) ID() string {
+	return "headerValueHelper"
+}
+
+func (h *headerValueHelper) HandleBuild(ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler) (out middleware.BuildOutput, metadata middleware.Metadata, err error) {
+	req, ok := in.Request.(*Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown transport type %T", in.Request)
+	}
+
+	for _, value := range h.headerValues {
+		if value.append {
+			req.Header.Add(value.header, value.value)
+		} else {
+			req.Header.Set(value.header, value.value)
+		}
+	}
+
+	return next.HandleBuild(ctx, in)
+}
+
+func getOrAddHeaderValueHelper(stack *middleware.Stack) (*headerValueHelper, error) {
+	id := (*headerValueHelper)(nil).ID()
```

Comment:
Is this casting or type asserting? We have used similar statements to assert the type, but not sure how this works here. 

Wouldn't this panic as headerValueHelper is nil?

Reply:
This would be initialization of a `nil` value of `headerValueHelper`. This is a way to call a method on a type, without initialing that type first. Since `ID` uses a pointer receiver, and does not use any member of `headerValueHelper` the `nil` type value will never be de-referenced. So no panic.

Reply:
Alternatives would be to create a variable of `headerValueHelper` and call ID separatly.

```go
var ider headerValueHelper

id := ider.ID()
```